infrastructure/smart-contracts/contracts/Achievements.sol

```diff
@@ -0,0 +1,147 @@
+            _tokenIdList.push(id);
+            tokenIdCount += 1;
+        }
+        totalSupply += 1;
+        totalSupplyById[id] += 1;
+
+        _mint(to, id, 1, data);
+    }
+
+    /**
+     * @dev Mints a batch of tokens to the recipient address. Only callable by MINT_ROLE holder.
+     * @param to Address to receive the newly minted tokens.
+     * @param ids Array of token IDs of tokens to mint.
+     * @param data Arbitrary data field.
+     */
+    function mintBatch(address to, uint256[] calldata ids, bytes memory data) public onlyMintRole {
+        require(ids.length > 0, "invalid ids param length");
+        
+        uint256[] memory amounts = new uint256[](ids.length);
+        for (uint256 i = 0; i < ids.length; i++) {
+            require(balanceOf(to, ids[i]) == 0, "address has non-zero token balance");
+            
+            if (!definedTokens[ids[i]]) {
+                _tokenIdList.push(ids[i]);
+                tokenIdCount += 1;
+            }
+            totalSupply += 1;
+            totalSupplyById[ids[i]] += 1;
+            amounts[i] = 1;
+        }
+
+        _mintBatch(to, ids, amounts, data);
+    }
+
+    /**
+     * @dev Gets an array of all defined token ids for this contract.
+     */
+    function getDefinedTokens() external view returns (uint256[] memory) {
+        return _tokenIdList;
+    }
+
+    /**
+     * @dev Gets a paginated array of defined token ids for this contract.
+     * @param pageStart Index for page start.
+     * @param pageEnd Index for page end.
+     */
+    function getDefinedTokens(uint256 pageStart, uint256 pageEnd) external view returns (uint256[] memory) {
+        require(pageStart < pageEnd, "invalid page bounds");
+
+        uint256 pageSize = pageEnd - pageStart;
+        uint256[] memory tokenIdPage = new uint256[](pageSize);
+        
+        uint256 cursor = pageStart;
+        for (uint256 i = 0; i < pageSize; i++) {
+            tokenIdPage[i] = _tokenIdList[cursor];
+            cursor++;
+        }
+        
+        return tokenIdPage;
+    }
+
+    function _safeTransferFrom(
+        address from,
+        address to,
+        uint256 id,
+        uint256 amount,
+        bytes memory data
+    ) internal pure override {
+        require(false, "not transferrable");
+    }
+
+    function _safeBatchTransferFrom(
+        address from,
+        address to,
+        uint256[] memory ids,
+        uint256[] memory amounts,
+        bytes memory data
+    ) internal pure override {
+        require(false, "not batch transferrable");
+    }
+
+    function _afterTokenTransfer(
+        address operator,
+        address from,
+        address to,
+        uint256[] memory ids,
+        uint256[] memory amounts,
+        bytes memory data
+    ) internal view override {
+        if (ids.length > 1) {
+            for (uint256 i = 0; i < ids.length; i++) {
+                require(balanceOf(to, ids[i]) == 1, "address has non-zero token balance");
```

Comment:
Note: this checks that multiple same token ids cannot be transferred via the `mintBatch()` function (e.g. where ids array looks like [1, 1, 1, 1] etc.)

Reply:
Do we need to overwrite `_afterTokenTransfer`, its not going to be triggered since the other transfers are reverting ?

Reply:
This require only throws when a user already owns the token before minting. It won't trigger with ids arrays like [1, 1] though because the minting doesn't occur until after the for loop, and the `_afterTokenTransfer()` logic catches that.